### PR TITLE
Benches/program mutualize

### PR DESCRIPTION
**Pull request by francois-caddet**
Use the same eval environment for all subbenches.
Also implement a new macro to have a clean description of benches

### CODE REVIEW

**Review by yannham — COMMENTED** [review]
Memory usage is now stable indeed! Looking good. Usually I'm reluctant to use macros without a good motivation (reducing painful boilerplate can be one). Here it seems it is mostly used to have named parameters (as opposed as using bare functions)? On the other hand, it just mimics `criterion_group` and isn't too involved, so I imagine it's fine.

File: utilities/src/lib.rs [hunk 1]
```diff
@@ -36,108 +39,212 @@ pub enum EvalMode {
...
-        c,
-    )
+pub struct Bench<'b> {
+    pub name: &'b str,
+    pub base_dir: &'b str,
+    pub subpath: &'b str,
+    pub subtest: Option<&'b str>,
+    pub args: Vec<String>,
+    pub eval_mode: EvalMode,
+    pub pred: Box<dyn Fn(Term) -> bool>,
```
[the remark this answers is not in the document]
Comment (francois-caddet) [reply utilities/src/lib.rs:49]: No, I don't think we actualy use it.

File: utilities/src/lib.rs [hunk 2]
```diff
@@ -36,108 +39,212 @@ pub enum EvalMode {
...
-        subtest,
-        args,
-        eval_mode,
-        |_| true,
-        c,
-    )
+pub fn bench_terms<'r>(rts: Vec<Bench<'r>>) -> Box<dyn Fn(&mut Criterion) + 'r> {
+    let mut cache = Cache::new();
+    let env = cache.prepare_stdlib().unwrap();
+    let eval_env = env.eval_env.clone();
```
[the remark this answers is not in the document]
Comment (francois-caddet) [reply utilities/src/lib.rs:159]: yes you're right.

**Comment by francois-caddet** [comment]
> Memory usage is now stable indeed! Looking good. Usually I'm reluctant to use macros without a good motivation (reducing painful boilerplate can be one). Here it seems it is mostly used to have named parameters (as opposed as using bare functions)? On the other hand, it just mimics `criterion_group` and isn't too involved, so I imagine it's fine.

Yes, the idea was to mimic `criterion::bench_group` because it can not be used if we want a "global" environment. Every function passed to `bench_group` is run independently. As in this PR we don't change to much the syntax but it's specific for nickel needs.

File: utilities/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (yannham) [reply utilities/src/lib.rs:121]: @francois-caddet It's still cloned in this version, is that on purpose?

[the remark this answers is not in the document]
Comment (yannham) [reply utilities/src/lib.rs:121]: Also, this code looks suspiciously identical to the one in the macro `ncl_bench_group`. Do we really need to keep the two versions?

File: benches/serialization/main.ncl [hunk 3]
```diff
@@ -0,0 +1,6 @@
+
+{
+    input.run = (
+        builtin.serialize `Json (import "input.json")
+    )
```
Comment (yannham) [inline benches/serialization/main.ncl:5]:
```suggestion
    input.run =
      import "input.json"
      |> builtin.serialize `Json
```

File: benches/serialization.rs [hunk 4]
```diff
@@ -1,25 +1,25 @@
...
-use nickel_lang::program::Program;
+use criterion::{criterion_main, Criterion};
+use nickel_lang_utilities::ncl_bench_group;
 use pprof::criterion::{Output, PProfProfiler};
-use std::io::Cursor;
+
 use std::path::PathBuf;
 
-fn round_trip(c: &mut Criterion) {
+fn serialization_path(subpath: &str) -> String {
```
Comment (yannham) [inline benches/serialization.rs:7]: I think this function is not used anymore (indeed, there is a compilation warning).

**Review by yannham — APPROVED** [review]
Left some remarks, but otherwise LGTM.

**Pull request merged**